src/history/historyTracker.ts

```diff
@@ -420,13 +424,45 @@ export class HistoryTracker {
     return this.currentHistoryStep.marks;
   }
 
+  /**
+   * Returns true if we need to get the entire document's text
+   * to process an individual change
+   */
+  private _isDocumentTextNeeded(): boolean {
+    // Get the current document's text length.
+    const documentEnd = new Position(0, 0).getDocumentEnd();
+    const newTextLength = this.vimState.editor.document.offsetAt(documentEnd);
+
+    const isTextLengthDiff = this.oldText.length !== newTextLength;
+
+    // Determine if we just switched modes.
+    // This prevents recording steps in between start-end of a historyStep.
+    const isModeDiff = this.currentMode !== this.vimState.currentMode;
+
+    const isNewHistoryStep =
```

Comment:
I think this logic might need to be a bit more complicated. I think a normal motion like `j` or `<C-d>` without any operator (`cj` for instance) should ideally never trigger a document diff, but currently it does. I'll have to do a bit more thinking about this, but maybe we keep track of the number of history steps and see if that's changed?

Reply:
Or maybe we just don't call `addChange` in the first place if it's a movement without an operator?

Reply:
Oh, we need to keep in mind that our definition of "movement" basically means able to be combined with an operator... maybe we just label certain commands as ones that never change document text.

Reply:
I'm going to play around with that idea and put up another PR. I think that might be an easy win that could work alongside this optimization.

Reply:
> maybe we just label certain commands as ones that never change document text
I like this idea, we definitely need a different solution to optimizing navigation.